cmake/compiler/gcc/target_warnings.cmake

```diff
@@ -82,9 +82,11 @@ endmacro()
 
 macro(toolchain_cc_warning_extended)
 
+if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "4.6.0")
   zephyr_cc_option(
```

Comment:
'zephyr_cc_option', as opposed to "zephyr_compile_options", is supposed to test that the compiler is compatible with the flag.

Any idea why this is not  working?

Perhaps cleaning the ToolchainCapabilityCache will change the result.

mv ~/.cache/zephyr/ToolchainCapabilityDatabase{,_backup}

Reply:
I think this might have been an issue with the particular build of gcc I am using.

I just checked this issue again and it is no longer complaining about -Wno-unused-but-set-variable; in fact, it is not complaining about -Wno-this-is-not-a-valid-warning either ...

I will look into this again, and if it is indeed a compiler issue, I will remove this commit.

Reply:
Great.

Reply:
Note that it is expected that there will be no complaints about -Wno-this-is-not-a-valid-warning because the build systems tests for toolchain compatibility and silently omits flags that are not supported.

Reply:
So it turns out the reason why I was getting `cc1: warning: unrecognized command line option "-Wno-unused-but-set-variable"` was that there was at least one other (not unrecognised command line option) warning displayed, and GCC only displays this warning if there are other warnings.

I created and compiled a test file called `test.c` with the following code:
`int main(void) { return 0; }`

As you can see below, GCC displays absolutely no warning- not even about the unrecognised command line option (despite having set the -Wall option).
```
stephanos@KONTRA:/mnt/d/MyDev/Repositories/zephyr-stephanos/samples/basic/blinky/build_tc32$ /opt/tc32/bin/tc32-elf-gcc -c -Wall -Wno-unused-but-set-variable test.c
```

Now, I remove `return 0;`, compile again and I get the following:
```
stephanos@KONTRA:/mnt/d/MyDev/Repositories/zephyr-stephanos/samples/basic/blinky/build_tc32$ /opt/tc32/bin/tc32-elf-gcc -c -Wall -Wno-unused-but-set-variable test.c
test.c: In function 'main':
test.c:3:1: warning: control reaches end of non-void function
At top level:
cc1: warning: unrecognized command line option "-Wno-unused-but-set-variable"
```

While I find that somewhat peculiar, that seems to be the standard behaviour because I tested again with the host GCC and it behaved in the same manner.

Reply:
So my understanding at this time is as follows:
* Dirty ToolchainCapabilityCache is not the problem.
* zephyr_cc_option is supposed to check if a warning is supported.
* CMake compiles a test file with `-Wno-unused-but-set-variable` that generates no other warning, so the unrecognised command line option warning is suppressed and it assumes that the warning is "supported."
* When there is no warning about anything in the project (which, of course, should always be the case), the unrecognised command line option warning is suppressed and, despite being unsupported, it goes unnoticed by the dev.
* But when there is at least one other warning in the project, that warning and the then-mysterious unrecognised command line option warning is displayed and confuses the dev.

My opinion on this is that this is less-than-ideal behaviour and should be corrected to prevent confusion (it is very likely that, as I did, a developer will try to correct the unrecognised command line option warning alongside another warning that is displayed).

One way to correct that problem would be to explicitly check the compiler version to see if a warning is supported before specifying it; a better way would be to get 'zephyr_cc_option' to properly check it.

Can you confirm if this issue is reproducible on your side (i.e. zephyr_cc_option failing to check and filter out the options that cause unrecognised command line option warning)?

Reply:
I can confirm everything you have found.

I wonder what the best approach is. If we can't find something better
a toolchain version check will be necessary.

Reply:
I see that we are not able to test for compatibility with -Wno-<warning> flags, but we are able to test for compatibility with -W<warning> flags.

I wonder if the cleanest solution would be for zephyr_cc_options to recognize that we are adding a -Wno-<warning> flag, and if so test the -W<warning> flag in its place and infer from that whether -Wno-<warning> is supported.

Reply:
PR posted: https://github.com/zephyrproject-rtos/zephyr/pull/18967

Reply:
I can confirm #18967 resolves this issue. I will remove commit 03dd4a81f4dc4ed39a7f006367ee7aab44888531 from this PR.